### PR TITLE
Add initial LoRA implementation

### PR DESCRIPTION
**Pull request by fchollet**

### CODE REVIEW

**Comment by fchollet** [comment]
@mattdangerw please take a look. Note that the `loralib` library does a few more things (scaling and dropout in particular) but it doesn't appear to be particularly useful. https://github.com/microsoft/LoRA/blob/main/loralib/layers.py

**Comment by fchollet** [comment]
Another note -- it is possible to reduce code redundancy via a LoRA base class. However if we only cover these 3 layers then it is likely not worth it.

**Comment by martin-gorner** [comment]
Question: how do you apply this LoRA implementation ? Presumably something like this?

```
decoder_block._self_attention_layer. _query_dense.enable_lora(rank=2)
```

Could be a more generic implementation. One that would check these boxes:
- it would not complexify the keras.layers.Dense implementation. We don't know if other LoRA-like techniques aren't on the horizon.
- It could easily be implemented by the user
- it could also be implemented as a helper in KerasNLP, without needing to change Keras core
- It could be applied without using private variables like _self_attention_layer or _query_dense, i.e. be applied through standard Keras introspection APIs

Quick and (very) incomplete proof of concept here:
https://colab.research.google.com/drive/1UrEJcFcIkHx4MVdg4ZXUx_CxB7gnpCrG?usp=sharing

**Comment by fchollet** [comment]
The issue with using `clone_model` is that it will only work for functional and sequential models (not subclassed models), and it won't work in-place. For instance, it won't work out of the box with the actual class that people use to build Transformers, the MHA layer (since it is a subclass that owns EinsumDense layers).

**Comment by martin-gorner** [comment]
I did notice the issue with clone_model. Another issue would be memory usage. This is not the solution, only a proof of concept of the UX.
The MHA layer issue is fixable: we could redesign it to be introspectable through regular Keras introspection APIs.
What we are missing is a generic mechanism for swapping or wrapping layers in a model.
Since maintaining a graph of layers is one of the key strengths of Keras, a swap/wrap API is a possible addition, and would be beneficial here and other similar cases.

**Comment by AakashKumarNain** [comment]
@fchollet I am aligned with @martin-gorner here. We now have `LoRA`, `QLoRA`, `ReLoRA` that are being used. Instead of making changes to the core layers, we ideally should have a `LoRALayer` where you have some abstract functions that needs to be implemented depending on the kind of layer subclassing it. There are at least two advantages I see with this approach:

1. The core layers remains unchanged. Even if LoRA becomes out of fashion after a year or so, we won't have to remove it again from core layers.
2. It becomes easy to quickly implement and add multiple variants of LoRA.

**Comment by fchollet** [comment]
I want to highlight the fact that it is always possible for you to invent new LoRA related techniques and implement them via layer subclasses that you insert in a model via `clone_model`. It's all completely viable. It's just not feature-complete (only works with Sequential/Functional models), so for the main API, we want something built-in, better-designed, that works with everything.

File: keras/layers/core/dense.py [hunk 1]
```diff
@@ -105,9 +118,18 @@ def build(self, input_shape):
             self.bias = None
         self.input_spec = InputSpec(min_ndim=2, axes={-1: input_dim})
         self.built = True
+        if self.lora_rank:
+            self.enable_lora(self.lora_rank)
 
     def call(self, inputs):
-        x = ops.matmul(inputs, self.kernel)
+        if self.lora_enabled:
+            kernel = self.kernel + ops.matmul(
```
Comment (fchollet) [inline keras/layers/core/dense.py:126]: Note: I get the feeling it will be better to have a `kernel` property that can be overridden.

**Comment by martin-gorner** [comment]
> implement them via layer subclasses that you insert in a model via clone_model
As you correctly pointed out above, clone_model does not work for LoRA-fying the dense layers in a TransformerBlock. You have to re-implement the entire TransformerBlock.

> LoRA, QLoRA, ReLoRA
And many more will be coming.

> only works with Sequential/Functional models
For subclassing models, model._my_private_layer = LoRAfy(model._my_private_layer) works fine, apart from the fact that you have to go through a private variable.
The problem with Keras today is that you cannot swap a layer for  its loRA equivalent using standard introspection APIs in the Sequential/Functional case.

**Comment by martin-gorner** [comment]
I looked at [ReLoRA](https://arxiv.org/pdf/2307.05695.pdf) and it seems to be a repeated application of LoRA. It should work with the `.enable_lora()` API that Francois has drafted, provided you can call enable/disable multiple times during training.

QLoRA can also be decomposed as Quantization + LoRA so it could theoretically work with the `.enable_lora()` API, but will depend on how we do quantization.

A Question mark remains about future layer replacement techniques.

**Comment by fchollet** [comment]
A thing I've done in the current PR to make it easier for future techniques to be supported is make `kernel` a property. It means you can create a custom LoRA layer by only overriding `enable_lora` and `kernel`, in a few lines of code. Then you can use the custom layer either standalone or as a layer replacement in `clone_model`. Again not feature complete but definitely good enough to start leveraging new techniques easily (though we need to think a bit about how to handle the MHA layer).

**Comment by AakashKumarNain** [comment]
> you can create a custom LoRA layer by only overriding enable_lora and kernel

Sounds good to me 🎉

File: keras/layers/core/einsum_dense.py [hunk 2]
```diff
@@ -1,5 +1,7 @@
 import re
 
+import numpy as np
```
Comment (mattdangerw) [inline keras/layers/core/einsum_dense.py:3]: Since einsums can be super general, I added some validation when I added this to KerasNLP.

https://github.com/mattdangerw/keras-nlp/blob/b8d86fe8fe0854b6fa98b7cfb84b73aafa5cd036/keras_nlp/layers/modeling/lora_dense.py#L22-L39

https://github.com/mattdangerw/keras-nlp/blob/b8d86fe8fe0854b6fa98b7cfb84b73aafa5cd036/keras_nlp/layers/modeling/lora_dense.py#L22-L39

We could consider doing something like this? Otherwise I am sure there could equation lines someone constructs that will not work with `enable_lora`.

**Review by mattdangerw — COMMENTED** [review]
Left some comments on the code itself.

I do get the worry about an overly complex Dense layer down the road (if we want more API similar to this), but no super compelling alternate design comes to mind to right now.

File: keras/layers/core/einsum_dense.py (hunk 2 above)
Comment (mattdangerw) [reply keras/layers/core/einsum_dense.py:3]: (these are links to the layer we have sense removed for this, but could still mimic the validation)

File: keras/layers/core/dense.py [hunk 3]
```diff
@@ -119,6 +146,54 @@ def compute_output_shape(self, input_shape):
...
+            raise ValueError(
+                "Lora is incompatible with kernel constraints. "
+                "In order to enable lora on this layer, remove the "
+                "`kernel_constraint` argument."
+            )
+        if not self.built:
+            raise ValueError(
+                "Cannot enable lora on a layer that isn't yet built."
+            )
+        self._tracker.locked = False
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras/layers/core/dense.py:160]: > This will basically invalidate any compiled function on a toplevel model right?

It would not. The behavior of a compiled `fit()` is that it sticks to the state of the model when it was last compiled. To get the new lorafied model, you need to recompile.

File: keras/layers/core/einsum_dense.py (hunk 2 above)
Comment (fchollet) [reply keras/layers/core/einsum_dense.py:3]: It will always work (unless the kernel is a scalar, but that's impossible), but a caveat is that we only factorize on the last axis, and for weird equations this may not be the desired behavior (it is clearly documented though). For all use cases that I know of, `axis=-1` would be the right behavior.

If you make the axis configurable, things get one order of magnitude more complex, so I didn't expose that option.

**Comment by mattdangerw** [comment]
Some general questions.

1) Do we want to expose a way to set the bias not trainable? This is actually the default for microsoft/LoRA and huggingface/peft (though keeping a trainable bias should be relatively cheap). See the note on bias here https://github.com/microsoft/LoRA#additional-notes

2) In general, what is our expected flow to freeze all weights in a model besides lora? This?
```python
for weight in model.weights:
    weight.trainable = "lora" in weight.name
```

3) Do we want a way to smush lora weights updates into the model without saving/re-loading the weights? E.g. to run maximally efficient inference in the process as training? Not necessary, just something to consider.
```python
model = build_model()
model.dense.enable_lora(4)
model.another_dense.enable_lora(4)
model.fit(...)
# Do we want to make this easier?
model.save_weights("merged_weights.keras.h5")
del model
model = build_model()
model.load_weights("merged_weights.keras.h5")
model.predict(...)
```

**Comment by fchollet** [comment]
> Do we want to expose a way to set the bias not trainable?

This is always doable via `layer.bias.trainable = False` (might need to check that is exists / is not `None`).
If this is the right way to do lora, then we should do it by default in `enable_lora()`. Is it actually worth exposing an option not to do it?

> In general, what is our expected flow to freeze all weights in a model besides lora? This?

The code snippet works -- or you could start by setting `w.trainable = False` on all weights (note: not the layers themselves), and *then* call `enable_lora()`.

> Do we want a way to smush lora weights updates into the model without saving/re-loading the weights? E.g. to run maximally efficient inference in the process as training? Not necessary, just something to consider.

We don't support for partial checkpoints so anything like this would have to be done manually (and you would have to roll out your own convention for lora partial checkpoints). Easy to do via `.assign()` on the lora weights of a lora-enabled model.

**Comment by martin-gorner** [comment]
> you can create a custom LoRA layer by only overriding enable_lora and kernel

That works. Can you then define the Lora-enabled Dense layer as a subclass of the Dense layer? Just trying to keep Dense a one-liner as much as possible.

> Another API we could consider that would be easier to extend at feature parity is an external def enable_lora(layer, ...)

I'd prefer we figure out a general layer replacement API:
* have a way to swap a layer for another one in the (functional) layer graph
* have an MHA/TransformerBlock implementation that makes their structure programmatically accessible (it's just 6 dense layers arranges in a certain way, there is nothing complex). I have a quick re-implementation [here](https://colab.research.google.com/drive/1UrEJcFcIkHx4MVdg4ZXUx_CxB7gnpCrG?usp=sharing#scrollTo=pHDnFYn8t8Id) to see what it could look like. 

In the end, the API would be very similar to your proposed `enable_lora(layer, ...)` but more generic, maybe even something like:
```model.traverse_layers().filter(
      # only EinsumDense layers
   ).map(
      # replace with LoraEinsumDense(original_layer)
)```

**Comment by mattdangerw** [comment]
I think going with what we have here sounds better over standalone external methods. A function that patches a bunch of methods on an existing object just feels a little overly complex and not super "keras-y." This has the nice property that the code is ready and simple. But I totally get the concern of polluting our precious dense layer and am rather torn.

I think the thing that feels missing to me right now is quantization, which feels very similar (override kernel, add state, affects saving). We should make sure our overall design feels consistent. Might make sense to revisit this once we have a quantization design as well.

File: keras/layers/core/dense.py [hunk 4]
```diff
@@ -105,6 +118,20 @@ def build(self, input_shape):
             self.bias = None
         self.input_spec = InputSpec(min_ndim=2, axes={-1: input_dim})
         self.built = True
+        if self.lora_rank:
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras/layers/core/dense.py:121]: We don't validate the value of `rank` so it may be 0 -- I think it's reasonable to interpret rank=0 as "no lora". Some APIs do this
